### DISCUSSION TITLE
Mutable State with Lifetime

### DISCUSSION THREAD

**Discussion by jcuenod** [tauri-apps/tauri#7180] [Q&A]
I'm trying to store the `Vec<Node<'_>>` that get's returned by simd_json's `to_tape` function in Tauri state. I'm new to Rust and Tauri, but if I understand correctly, the reason for the lifetime is that `to_tape` takes a `&mut [u8]` and avoids allocations until necessary, so access to the slice is necessary as long as the tape is accessible. That is, I want an `open` command that loads the tape and a `query` command that does particular lookups.

I've made a number of attempts with Mutex and RwLock etc., but the slice does not live long enough. How should I be doing this? If this is non-idiomatic, let me know.

**Comment by FabianLars** [comment]
**I'd love to be proven wrong but i feel like the only sane solution is to use the [BorrowedValue api](https://docs.rs/simd-json/latest/simd_json/#values-api) which ofc is incompatible with the "avoids allocations until necessary" part :/**

The only alternative that comes to mind is using a self-referential type which is pretty unsafe and probably easy to mess up, so imho not worth it. Also i think this can only work if you own the data `&mut [u8]` borrows, so depending on where you get the data from it's probably useless.
For what it's worth, i thought this guide was pretty good when i looked into it a while ago https://morestina.net/blog/1868/self-referential-types-for-fun-and-profit#Implementation_of_a_self-referential_struct. A relatively simple example may look like this:
```rs
struct TapeWrapper {
    tape: Vec<simd_json::Node<'static>>,
    data: aliasable::boxed::AliasableBox<Vec<u8>>,
}

// somewhere you have access to App/AppHandle/Window
let mut d = aliasable::boxed::AliasableBox::from_unique(Box::new(
    br#"{"some": ["key", "value", 2]}"#.to_vec(),
));
let v = simd_json::to_tape(&mut d).unwrap();
let v = unsafe { std::mem::transmute(v) };

let x = TapeWrapper { tape: v, data: d };

app.manage(x);
```

I really hope i'm just missing an obvious solution here 😅

**Comment by jcuenod** [comment]
Yes, I'm also hoping to avoid `unsafe` if possible.
